File: AllToDo-Android/app/src/main/java/com/example/alltodo/utils/SmartLocationManager.kt
package com.example.alltodo.utils

import android.location.Location
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.pow

object SmartLocationManager {
    // Precision: 100,000 => 5 decimal places => ~1.1 meters resolution
    private const val PRECISION = 100000.0
    private const val BASE_ZOOM = 18f
    // At Zoom 18, we want 10m threshold. 
    // 10m in degrees (Lat) ~= 10 / 111000 = 0.00009 deg
    // In Integer units (x 100,000) => 9 units.
    private const val BASE_THRESHOLD_UNITS = 9 

    data class IntLocation(val lat: Int, val lon: Int)

    fun toIntLocation(loc: Location): IntLocation {
        return IntLocation(
            (loc.latitude * PRECISION).toInt(),
            (loc.longitude * PRECISION).toInt()
        )
    }

    /**
     * Determines if the map/location should be updated based on distance moved and zoom level.
     * Rule: Update if moved > 10m (at Zoom 18).
     * Threshold scales with Zoom: Lower Zoom (Far) -> Larger Threshold.
     * Formula: Threshold = 10m * 2^(18 - Zoom)
     */
    fun shouldUpdate(lastLoc: IntLocation?, newLoc: Location, currentZoom: Float): Boolean {
        if (lastLoc == null) return true

        val newIntLoc = toIntLocation(newLoc)
        val deltaLat = abs(lastLoc.lat - newIntLoc.lat)
        val deltaLon = abs(lastLoc.lon - newIntLoc.lon)

        // Calculate dynamic threshold
        // e.g. Zoom 18 -> Factor 1 -> 9 units (10m)
        // e.g. Zoom 15 -> Factor 8 -> 72 units (80m)
        // e.g. Zoom 20 -> Factor 0.25 -> 2 units (2.2m)
        // We cap max calculation at generous bounds to prevent overflow or weirdness
        val factor = 2.0.pow((BASE_ZOOM - currentZoom).toDouble())
        val threshold = (BASE_THRESHOLD_UNITS * factor).toInt().coerceAtLeast(2) // Min 2 units (~2m)
        
        // Simple Box Check (Manhattan-ish) for speed
        return deltaLat > threshold || deltaLon > threshold
    }

    /**
     * Checks if distance is > 500km using simplified integer math.
     * 500km = 500,000m.
     * Units = 500,000m / 1.11m ~= 450,000 units.
     */
    fun isFar(loc1: Location, loc2: Location): Boolean {
        val p1 = toIntLocation(loc1)
        val p2 = toIntLocation(loc2)
        
        val dy = (p1.lat - p2.lat).toLong()
        
        // Correct longitude for latitude shrinking
        val avgLatRad = Math.toRadians((loc1.latitude + loc2.latitude) / 2.0)
        val dx = ((p1.lon - p2.lon) * cos(avgLatRad)).toLong()
        
        val distSq = dx*dx + dy*dy
        val limit = 450000L // ~500km in units
        return distSq > (limit * limit)
    }
}

File: AllToDo-iOS/AllToDo/Utilities/SmartLocationManager.swift
import Foundation
import CoreLocation

class SmartLocationManager {
    static let shared = SmartLocationManager()
    
    // Precision: 100,000 => ~1.1m resolution
    private let PRECISION: Double = 100000.0
    
    typealias IntLocation = (lat: Int, lon: Int)
    
    func toIntLocation(_ loc: CLLocation) -> IntLocation {
        return (
            Int(loc.coordinate.latitude * PRECISION),
            Int(loc.coordinate.longitude * PRECISION)
        )
    }
    
    /// Determines if location update is needed based on movement and map span (zoom).
    /// - Parameters:
    ///   - lastLoc: Last processed location in integer units
    ///   - newLoc: New GPS location
    ///   - currentSpan: Map latitude span (degrees). Smaller span = higher zoom.
    /// - Returns: True if update is significant enough to redraw.
    func shouldUpdate(lastLoc: IntLocation?, newLoc: CLLocation, currentSpan: Double) -> Bool {
        guard let last = lastLoc else { return true }
        
        let newInt = toIntLocation(newLoc)
        let deltaLat = abs(last.lat - newInt.lat)
        let deltaLon = abs(last.lon - newInt.lon)
        
        // Threshold Logic:
        // We want sensitivity to be roughly 1/500th of the screen height.
        // e.g. Span 0.005 (Zoom ~17) -> Threshold 0.00001 deg (~1.1m) -> 1 unit.
        // e.g. Span 0.5 (Zoom ~10)   -> Threshold 0.001 deg (~100m)   -> 100 units.
        // Formula: Threshold Units = (Span * 200).
        // Clamped to min 2 units (~2m) to avoid jitter at max zoom.
        
        let calculatedThreshold = Int(currentSpan * 200.0)
        let threshold = max(2, calculatedThreshold)
        
        return deltaLat > threshold || deltaLon > threshold
    }
    
    /// Checks if distance > 500km using integer math.
    func isFar(_ loc1: CLLocation, _ loc2: CLLocation) -> Bool {
        let p1 = toIntLocation(loc1)
        let p2 = toIntLocation(loc2)
        return isFar(lat1: p1.lat, lon1: p1.lon, lat2: p2.lat, lon2: p2.lon)
    }
    
    /// Checks if distance > 500km using pure integer coordinates (no double conversion for inputs).
    /// - Parameters:
    ///   - lat1: Latitude * 100,000
    ///   - lon1: Longitude * 100,000
    ///   - lat2: Latitude * 100,000
    ///   - lon2: Longitude * 100,000
    func isFar(lat1: Int, lon1: Int, lat2: Int, lon2: Int) -> Bool {
        let dy = Int64(lat1 - lat2)
        
        // Longitude distance depends on latitude (cos(avgLat)).
        // We still need Double for Cosine calculation unless we use a lookup table,
        // but inputs are Ints.
        // Approx Avg Lat in Rad
        let avgLatRad = Double(lat1 + lat2) / 2.0 / Double(PRECISION) * .pi / 180.0
        let dx = Int64(Double(lon1 - lon2) * cos(avgLatRad))
        
        let distSq = dx*dx + dy*dy
        
        // 500km Limit (Reverted to User Standard)
        // 500km ~= 4.5 deg lat.
        // 4.5 deg * 100,000 = 450,000 units.
        let limit: Int64 = 450000 
        return distSq > (limit * limit)
    }
}
